{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os.path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import hdbscan\n",
    "import time\n",
    "from sklearn import metrics\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics.pairwise import euclidean_distances\n",
    "\n",
    "import re #regular expression matching for removing unwanted columns by name \n",
    "import natsort as ns #3rd party package for natural sorting "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def raw_data_cleanup(filename):\n",
    "    \n",
    "    \"\"\"\n",
    "    Imports RNAseq .csv file and does basic clean up of \"FM40\" \n",
    "        -sorts FM40 timecourse sequence chronologically\n",
    "        -removes all QC data and non FM40 columns\n",
    "        -returns dataframe with locus tag set as index\n",
    "    \"\"\"\n",
    "\n",
    "    if os.path.isfile(filename):\n",
    "        print(\"{} was located in the directory\".format(filename))\n",
    "                \n",
    "#import the data\n",
    "        data0_raw = pd.read_csv(filename, sep = \"\\t\") \n",
    "        print(\"{} was imported into dataframe\".format(filename))\n",
    "        \n",
    "#removing all QC data\n",
    "        data1_noQC = data0_raw.select(lambda x: not re.search(\"QC\", x), axis = 1) \n",
    "        print(\"QC columns were removed from dataframe\")\n",
    "\n",
    "#removing all non FM40 data\n",
    "        data2_FM40only = data1_noQC.select(lambda x: re.search(\"FM40\", x), axis = 1)\n",
    "        print(\"All non FM40 data were removed from dataframe\")\n",
    "        \n",
    "#naturally sorting FM40 data by columns \n",
    "        cols = list(ns.natsorted(data2_FM40only.columns))\n",
    "        data3_sorted=data2_FM40only[cols]\n",
    "        print(\"All FM40 columns were sorted by timecourse sequence\")\n",
    "        \n",
    "#adding the descriptor columns back to FM40\n",
    "        qualitative = data0_raw.loc[:,\"locus_tag\":\"translation\"]\n",
    "        data4_sorted = pd.concat([qualitative, data3_sorted], axis = 1)\n",
    "\n",
    "#setting locus tag to be the index\n",
    "        data5_index = data4_sorted.set_index(\"locus_tag\")\n",
    "        \n",
    "    \n",
    "        print(\"Clean-up of raw data complete\")\n",
    "        return data5_index\n",
    "        \n",
    "    else:\n",
    "        print(\"{} does not exist in directory. Function was not complete.\".format(filename))\n",
    "        return\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def TPM_counts(dataframe, \n",
    "              gene_start,\n",
    "              gene_stop,\n",
    "              columns,\n",
    "              remove_zero = True):\n",
    "    \n",
    "    \"\"\"\n",
    "    TPM_counts(dataframe, gene_start, gene_stop, columns):\n",
    "\n",
    "    returns a dataframe with TPM instead of reads\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    daraframe = dataframe object variable\n",
    "    gene_start = string with column name containing gene start coordinate\n",
    "    gene_stop = string with column name containing gene stop coordinate\n",
    "    columns = list of strings of column names to be converted to TPM\n",
    "    remove_zero = if True, will remove rows containing zero expression\n",
    "\n",
    "\n",
    "    Run the following two lines to properly execute this function:\n",
    "\n",
    "    columns = ['5GB1_FM40_T0m_TR2', '5GB1_FM40_T10m_TR3', '5GB1_FM40_T20m_TR2', '5GB1_FM40_T40m_TR1',\n",
    "           '5GB1_FM40_T60m_TR1', '5GB1_FM40_T90m_TR2', '5GB1_FM40_T150m_TR1_remake', '5GB1_FM40_T180m_TR1']\n",
    "\n",
    "    TPM_counts(df,\"start_coord\",\"end_coord\",columns)\n",
    "    \"\"\"\n",
    "    \n",
    "    #create empty dataframe \n",
    "    gene_length = pd.DataFrame()\n",
    "    \n",
    "    #gene length in kilo base pairs as new column\n",
    "    gene_length[\"gene_length\"] = (dataframe[gene_stop]- dataframe[gene_start] + 1)/1000   \n",
    "    \n",
    "    #normalize read counts by gene length in kilo base pairs\n",
    "    RPK = dataframe.loc[:,columns].div(gene_length.gene_length, axis=0) \n",
    "    \n",
    "    #creating a series with the sums of each FM40 column / 1,000,000\n",
    "    norm_sum = RPK.sum(axis=0)/1000000 \n",
    "    norm_sum1 = pd.Series.to_frame(norm_sum)\n",
    "    norm_sum2 = norm_sum1.T\n",
    "    \n",
    "    #dividing by the the total transcript counts in each repicate\n",
    "    TPM = RPK.div(norm_sum2.ix[0]) \n",
    "    \n",
    "    dataframe.loc[:,columns] = TPM\n",
    "    \n",
    "    if remove_zero:\n",
    "        dataframe_values = dataframe.loc[:,columns] \n",
    "        remove_index = dataframe_values[dataframe_values.isin([0]).any(axis=1)].index\n",
    "        \n",
    "        dataframe = dataframe.drop(remove_index)\n",
    "\n",
    "    return dataframe "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Log2 fold transfrom of the TPM data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def log_2_transform(dataframe,\n",
    "                    first_data_column,\n",
    "                    last_data_column):\n",
    "                  \n",
    "    \"\"\"\n",
    "    log_2_transform(dataframe, \n",
    "                    first_data_column, \n",
    "                    last_data_column)\n",
    "    \n",
    "    Return a new dataframe with the range of data columns log2 transformed. \n",
    "    *all zero values are changed to 1 (yield 0 after transform)\n",
    "    *all values less than 1 are changed to 1 (yield 0 after transform)\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    daraframe = dataframe object variable\n",
    "    first_data_column = first column that contains actual data (first non categorical)\n",
    "    last_data_column = last column taht contains actual data (last non categorigal column)\n",
    "\n",
    "    Run the following to execute the function for Cu transition dataset. \n",
    "\n",
    "    log_2_transform(df, \"5GB1_FM40_T0m_TR2\", \"5GB1_FM40_T180m_TR1\") \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    df_data = dataframe.loc[:,first_data_column:last_data_column] #isolate the data\n",
    "    \n",
    "    df_data = df_data.replace(0,1) #replace all zeros with 1s\n",
    "    \n",
    "    df_data[df_data<1] = 1 #replace all values less than 1 with 1\n",
    "    \n",
    "    df_data_log2 = df_data.apply(np.log2)\n",
    "    \n",
    "    return df_data_log2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mean center the data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def mean_center(df, first_data_column, last_data_column):       \n",
    "        \n",
    "        \n",
    "    \"\"\"\n",
    "    mean_center(dataframe, \n",
    "                first_data_column, \n",
    "                last_data_column)\n",
    "    \n",
    "    Return a new dataframe with the range of data columns log2 transformed. \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    daraframe = dataframe object variable\n",
    "    first_data_column = first column that contains actual data (first non categorical)\n",
    "    last_data_column = last column taht contains actual data (last non categorigal column)\n",
    "\n",
    "    Run the following to execute the function for Cu transition dataset. \n",
    "\n",
    "    mean_center(df, \"5GB1_FM40_T0m_TR2\", \"5GB1_FM40_T180m_TR1\") \n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    df2_TPM_values = df2_TPM.loc[:,first_data_column:last_data_column] #isolating the data values \n",
    "    df2_TPM_values_T = df2_TPM_values.T #transposing the data\n",
    "\n",
    "    standard_scaler = StandardScaler(with_std=False)\n",
    "    TPM_counts_mean_centered = standard_scaler.fit_transform(df2_TPM_values_T) #mean centering the data \n",
    "\n",
    "    TPM_counts_mean_centered = pd.DataFrame(TPM_counts_mean_centered) #back to Dataframe\n",
    "\n",
    "    #transposing back to original form and reincerting indeces and columns \n",
    "    my_index = df2_TPM_values.index\n",
    "    my_columns = df2_TPM_values.columns\n",
    "\n",
    "    TPM_counts_mean_centered = TPM_counts_mean_centered.T\n",
    "    TPM_counts_mean_centered.set_index(my_index, inplace=True)\n",
    "    TPM_counts_mean_centered.columns = my_columns\n",
    "    \n",
    "    return TPM_counts_mean_centered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pariwise distance metric table - euclidean distance "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def euclidean_distance(dataframe, first_data_column, last_data_column):\n",
    "    \n",
    "    \"\"\"\n",
    "    euclidean_distance(dataframe, \n",
    "                first_data_column, \n",
    "                last_data_column)\n",
    "    \n",
    "    Return a new dataframe - pairwise distance metric table, euclidean distance between every pair of rows. \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    daraframe = dataframe object variable\n",
    "    first_data_column = first column that contains actual data (first non categorical)\n",
    "    last_data_column = last column taht contains actual data (last non categorigal column)\n",
    "\n",
    "    Run the following to execute the function for Cu transition dataset. \n",
    "\n",
    "    euclidean_distance(df, \"5GB1_FM40_T0m_TR2\", \"5GB1_FM40_T180m_TR1\") \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    df_values = dataframe.loc[:,first_data_column:last_data_column] #isolating the data values \n",
    "\n",
    "    df_euclidean_distance = pd.DataFrame(euclidean_distances(df_values))\n",
    "\n",
    "    my_index = dataframe.index\n",
    "    \n",
    "    df_euclidean_distance = df_euclidean_distance.set_index(my_index)\n",
    "    df_euclidean_distance.columns = my_index\n",
    "    \n",
    "    return df_euclidean_distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def congruency_table(df, \n",
    "         data_clm_strt, \n",
    "         data_clm_stop, \n",
    "         step, \n",
    "         mask_diagonal=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    corr(df, data_clm_strt, data_clm_stop, step = len(df.columns), mask_diagonal=False)\n",
    "    \n",
    "    returns a new datafram - congruency table - a pairwise pearson correlation matrix for every row pair\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    df - dataframe argument - recommended to use TPM counts for RNAseq datasets. \n",
    "    data_clm_strt = first column that contains data to be processed \n",
    "    data_clm_stop = last column that contains data to be processed\n",
    "    step = length of dataset \n",
    "    mask_diagonal = mask diagonal values which shoud come out as 1\n",
    "    \n",
    "    \n",
    "    Run the following lines to execute the function for my data\n",
    "    \n",
    "    congruency_table(df1, \"5GB1_FM40_T0m_TR2\" , \"5GB1_FM40_T180m_TR1\")\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    df = df.loc[:, data_clm_strt: data_clm_stop] #isolating the rows that are relavent to us. \n",
    "    df = df.T \n",
    "    \n",
    "    n = df.shape[0]\n",
    "\n",
    "    def corr_closure(df):\n",
    "        d = df.values\n",
    "        sums = d.sum(0, keepdims=True)\n",
    "        stds = d.std(0, keepdims=True)\n",
    "\n",
    "        def corr_(k=0, l=10):\n",
    "            d2 = d.T.dot(d[:, k:l])\n",
    "            sums2 = sums.T.dot(sums[:, k:l])\n",
    "            stds2 = stds.T.dot(stds[:, k:l])\n",
    "\n",
    "            return pd.DataFrame((d2 - sums2 / n) / stds2 / n,\n",
    "                                df.columns, df.columns[k:l])\n",
    "\n",
    "        return corr_\n",
    "\n",
    "    c = corr_closure(df)\n",
    "\n",
    "    step = min(step, df.shape[1])\n",
    "\n",
    "    tups = zip(range(0, n, step), range(step, n + step, step))\n",
    "\n",
    "    corr_table = pd.concat([c(*t) for t in tups], axis=1)\n",
    "\n",
    "    corr_table = corr_table.fillna(value = 0) #there are 94 NA values - filling them with zeros \n",
    "    \n",
    "    if mask_diagonal:\n",
    "        np.fill_diagonal(corr_table.values, np.nan)\n",
    "\n",
    "    return corr_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5G_counts.tsv was located in the directory\n",
      "5G_counts.tsv was imported into dataframe\n",
      "QC columns were removed from dataframe\n",
      "All non FM40 data were removed from dataframe\n",
      "All FM40 columns were sorted by timecourse sequence\n",
      "Clean-up of raw data complete\n",
      "The shape of the TPM table is  (4480, 16)\n",
      "The shape of the pearson_r matrix is  (4480, 4480)\n"
     ]
    }
   ],
   "source": [
    "df1_raw_FM40 = raw_data_cleanup(\"5G_counts.tsv\")\n",
    "\n",
    "\n",
    "columns = ['5GB1_FM40_T0m_TR2', '5GB1_FM40_T10m_TR3', '5GB1_FM40_T20m_TR2', '5GB1_FM40_T40m_TR1',\n",
    "           '5GB1_FM40_T60m_TR1', '5GB1_FM40_T90m_TR2', '5GB1_FM40_T150m_TR1_remake', '5GB1_FM40_T180m_TR1']\n",
    "\n",
    "df2_TPM = TPM_counts(df1_raw_FM40, \"start_coord\", \"end_coord\",columns, remove_zero = True)  #TPM counts\n",
    "df2_TPM_log2 = log_2_transform(df2_TPM, \"5GB1_FM40_T0m_TR2\",\"5GB1_FM40_T180m_TR1\") #TPM log 2 transformed \n",
    "df2_TPM_mean = mean_center(df2_TPM, \"5GB1_FM40_T0m_TR2\",\"5GB1_FM40_T180m_TR1\") #TPM mean centered \n",
    "\n",
    "df3_pearson_r = congruency_table(df2_TPM, \"5GB1_FM40_T0m_TR2\" , \"5GB1_FM40_T180m_TR1\", step = df2_TPM.shape[0])\n",
    "df3_euclidean_mean = euclidean_distance(df2_TPM_mean, \"5GB1_FM40_T0m_TR2\" , \"5GB1_FM40_T180m_TR1\")\n",
    "df3_euclidean_log2 = euclidean_distance(df2_TPM_mean, \"5GB1_FM40_T0m_TR2\" , \"5GB1_FM40_T180m_TR1\" )\n",
    "\n",
    "print(\"The shape of the TPM table is \", df2_TPM.shape)\n",
    "print(\"The shape of the pearson_r matrix is \", df3_pearson_r.shape)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "   ### Clustering pearsons_r with HDBSCAN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "time to cluster 4.091894865036011\n"
     ]
    }
   ],
   "source": [
    "# Clustering the pearsons_R with N/A vlaues removed \n",
    "\n",
    "hdb_t1 = time.time()\n",
    "hdb_pearson_r = hdbscan.HDBSCAN(metric = \"precomputed\", min_cluster_size=10).fit(df3_pearson_r)\n",
    "hdb_pearson_r_labels = hdb_pearson_r.labels_\n",
    "hdb_elapsed_time = time.time() - hdb_t1\n",
    "print(\"time to cluster\", hdb_elapsed_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[-1  0  1]\n",
      "[224  15]\n"
     ]
    }
   ],
   "source": [
    "print(np.unique(hdb_pearson_r_labels)) # unique bins, zero is noise\n",
    "print(np.bincount(hdb_pearson_r_labels[hdb_pearson_r_labels!=-1])) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pearson_clusters = {i: np.where(hdb_pearson_r_labels == i)[0] for i in range(2)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>product</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>locus</th>\n",
       "      <th>start_coord</th>\n",
       "      <th>end_coord</th>\n",
       "      <th>note</th>\n",
       "      <th>translation</th>\n",
       "      <th>5GB1_FM40_T0m_TR2</th>\n",
       "      <th>5GB1_FM40_T10m_TR3</th>\n",
       "      <th>5GB1_FM40_T20m_TR2</th>\n",
       "      <th>5GB1_FM40_T40m_TR1</th>\n",
       "      <th>5GB1_FM40_T60m_TR1</th>\n",
       "      <th>5GB1_FM40_T90m_TR2</th>\n",
       "      <th>5GB1_FM40_T150m_TR1_remake</th>\n",
       "      <th>5GB1_FM40_T180m_TR1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MBURv2_160283</th>\n",
       "      <td>putative ferric siderophore transport system, ...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>3576162</td>\n",
       "      <td>3576686</td>\n",
       "      <td>Evidence 3 : Function proposed based on presen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>118.243568</td>\n",
       "      <td>112.275736</td>\n",
       "      <td>110.964306</td>\n",
       "      <td>100.836391</td>\n",
       "      <td>77.197657</td>\n",
       "      <td>70.032566</td>\n",
       "      <td>73.893920</td>\n",
       "      <td>95.436505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_20180</th>\n",
       "      <td>ferrous iron transporter, protein A</td>\n",
       "      <td>CDS</td>\n",
       "      <td>feoA</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>281921</td>\n",
       "      <td>282154</td>\n",
       "      <td>Evidence 2a : Function of homologous gene expe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>210.424105</td>\n",
       "      <td>158.632845</td>\n",
       "      <td>181.080545</td>\n",
       "      <td>126.342956</td>\n",
       "      <td>61.051857</td>\n",
       "      <td>70.194605</td>\n",
       "      <td>95.464947</td>\n",
       "      <td>118.980515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_20181</th>\n",
       "      <td>fused ferrous iron transporter, protein B: GTP...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>feoB</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>282154</td>\n",
       "      <td>284478</td>\n",
       "      <td>Evidence 2a : Function of homologous gene expe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.819838</td>\n",
       "      <td>86.744847</td>\n",
       "      <td>91.146440</td>\n",
       "      <td>71.621996</td>\n",
       "      <td>32.461622</td>\n",
       "      <td>39.674872</td>\n",
       "      <td>55.479556</td>\n",
       "      <td>64.157192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_20183</th>\n",
       "      <td>conserved protein of unknown function</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>284953</td>\n",
       "      <td>286617</td>\n",
       "      <td>Evidence 4 : Homologs of previously reported g...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>92.724781</td>\n",
       "      <td>77.558180</td>\n",
       "      <td>86.194789</td>\n",
       "      <td>70.352941</td>\n",
       "      <td>49.814896</td>\n",
       "      <td>47.463966</td>\n",
       "      <td>55.260844</td>\n",
       "      <td>59.371024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_210346</th>\n",
       "      <td>Histidine kinase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>4474339</td>\n",
       "      <td>4475010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70.518396</td>\n",
       "      <td>68.822867</td>\n",
       "      <td>66.867096</td>\n",
       "      <td>63.493064</td>\n",
       "      <td>49.910520</td>\n",
       "      <td>54.874815</td>\n",
       "      <td>53.845876</td>\n",
       "      <td>59.197819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_240071</th>\n",
       "      <td>conserved protein of unknown function</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>4822700</td>\n",
       "      <td>4823671</td>\n",
       "      <td>Evidence 4 : Homologs of previously reported g...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>115.521215</td>\n",
       "      <td>125.266171</td>\n",
       "      <td>122.873477</td>\n",
       "      <td>171.859978</td>\n",
       "      <td>242.070964</td>\n",
       "      <td>230.034969</td>\n",
       "      <td>215.806091</td>\n",
       "      <td>173.416287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_240072</th>\n",
       "      <td>conserved protein of unknown function</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>4823671</td>\n",
       "      <td>4825122</td>\n",
       "      <td>Evidence 4 : Homologs of previously reported g...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>152.884221</td>\n",
       "      <td>166.004637</td>\n",
       "      <td>159.391668</td>\n",
       "      <td>244.378006</td>\n",
       "      <td>339.549453</td>\n",
       "      <td>333.040188</td>\n",
       "      <td>298.420330</td>\n",
       "      <td>260.795367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_30040</th>\n",
       "      <td>conserved exported protein of unknown function</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>648480</td>\n",
       "      <td>649751</td>\n",
       "      <td>Evidence 4 : Homologs of previously reported g...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>346.196665</td>\n",
       "      <td>456.748233</td>\n",
       "      <td>416.782502</td>\n",
       "      <td>555.641953</td>\n",
       "      <td>1007.956458</td>\n",
       "      <td>965.279974</td>\n",
       "      <td>855.040811</td>\n",
       "      <td>766.323303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_30115</th>\n",
       "      <td>Integrase/recombinase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>int</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>749393</td>\n",
       "      <td>750415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.954208</td>\n",
       "      <td>19.738377</td>\n",
       "      <td>18.230891</td>\n",
       "      <td>15.962341</td>\n",
       "      <td>11.687684</td>\n",
       "      <td>12.175099</td>\n",
       "      <td>14.009989</td>\n",
       "      <td>14.678014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_60111</th>\n",
       "      <td>nitrite reductase, large subunit, NAD(P)H-binding</td>\n",
       "      <td>CDS</td>\n",
       "      <td>nirB</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>1491766</td>\n",
       "      <td>1494321</td>\n",
       "      <td>Evidence 2a : Function of homologous gene expe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>605.258423</td>\n",
       "      <td>691.437830</td>\n",
       "      <td>667.622137</td>\n",
       "      <td>817.519655</td>\n",
       "      <td>1077.573416</td>\n",
       "      <td>1024.202207</td>\n",
       "      <td>988.430960</td>\n",
       "      <td>914.548557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_60113</th>\n",
       "      <td>Response regulator receiver and ANTAR domain p...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>1496503</td>\n",
       "      <td>1497081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.493069</td>\n",
       "      <td>27.461091</td>\n",
       "      <td>27.963444</td>\n",
       "      <td>34.230125</td>\n",
       "      <td>49.110936</td>\n",
       "      <td>51.477168</td>\n",
       "      <td>44.846438</td>\n",
       "      <td>43.132774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_60177</th>\n",
       "      <td>Transcriptional antiterminator, Rof</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>1561635</td>\n",
       "      <td>1561889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>245.862088</td>\n",
       "      <td>217.642197</td>\n",
       "      <td>254.978512</td>\n",
       "      <td>131.689317</td>\n",
       "      <td>45.544737</td>\n",
       "      <td>44.577813</td>\n",
       "      <td>81.965303</td>\n",
       "      <td>131.877279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_80064</th>\n",
       "      <td>fused DNA-binding response regulator in two-co...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>glnG</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>1901015</td>\n",
       "      <td>1902424</td>\n",
       "      <td>Evidence 2a : Function of homologous gene expe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>108.514855</td>\n",
       "      <td>116.924797</td>\n",
       "      <td>121.151452</td>\n",
       "      <td>163.911171</td>\n",
       "      <td>229.124357</td>\n",
       "      <td>225.811603</td>\n",
       "      <td>183.639002</td>\n",
       "      <td>180.891342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_80067</th>\n",
       "      <td>nitrogen assimilation regulatory protein for G...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>glnK</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>1906036</td>\n",
       "      <td>1906374</td>\n",
       "      <td>Evidence 2a : Function of homologous gene expe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1236.258838</td>\n",
       "      <td>1480.016871</td>\n",
       "      <td>1362.990010</td>\n",
       "      <td>2351.761430</td>\n",
       "      <td>3215.524758</td>\n",
       "      <td>3305.708022</td>\n",
       "      <td>3031.353754</td>\n",
       "      <td>2351.879330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MBURv2_80070</th>\n",
       "      <td>ammonium transporter</td>\n",
       "      <td>CDS</td>\n",
       "      <td>amtB</td>\n",
       "      <td>MBURv2</td>\n",
       "      <td>1908452</td>\n",
       "      <td>1909744</td>\n",
       "      <td>Evidence 2a : Function of homologous gene expe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>178.796228</td>\n",
       "      <td>216.015070</td>\n",
       "      <td>211.287476</td>\n",
       "      <td>293.525608</td>\n",
       "      <td>434.560734</td>\n",
       "      <td>446.470951</td>\n",
       "      <td>376.033774</td>\n",
       "      <td>368.793380</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         product type  \\\n",
       "locus_tag                                                               \n",
       "MBURv2_160283  putative ferric siderophore transport system, ...  CDS   \n",
       "MBURv2_20180                 ferrous iron transporter, protein A  CDS   \n",
       "MBURv2_20181   fused ferrous iron transporter, protein B: GTP...  CDS   \n",
       "MBURv2_20183               conserved protein of unknown function  CDS   \n",
       "MBURv2_210346                                   Histidine kinase  CDS   \n",
       "MBURv2_240071              conserved protein of unknown function  CDS   \n",
       "MBURv2_240072              conserved protein of unknown function  CDS   \n",
       "MBURv2_30040      conserved exported protein of unknown function  CDS   \n",
       "MBURv2_30115                               Integrase/recombinase  CDS   \n",
       "MBURv2_60111   nitrite reductase, large subunit, NAD(P)H-binding  CDS   \n",
       "MBURv2_60113   Response regulator receiver and ANTAR domain p...  CDS   \n",
       "MBURv2_60177                 Transcriptional antiterminator, Rof  CDS   \n",
       "MBURv2_80064   fused DNA-binding response regulator in two-co...  CDS   \n",
       "MBURv2_80067   nitrogen assimilation regulatory protein for G...  CDS   \n",
       "MBURv2_80070                                ammonium transporter  CDS   \n",
       "\n",
       "              gene_symbol   locus  start_coord  end_coord  \\\n",
       "locus_tag                                                   \n",
       "MBURv2_160283         NaN  MBURv2      3576162    3576686   \n",
       "MBURv2_20180         feoA  MBURv2       281921     282154   \n",
       "MBURv2_20181         feoB  MBURv2       282154     284478   \n",
       "MBURv2_20183          NaN  MBURv2       284953     286617   \n",
       "MBURv2_210346         NaN  MBURv2      4474339    4475010   \n",
       "MBURv2_240071         NaN  MBURv2      4822700    4823671   \n",
       "MBURv2_240072         NaN  MBURv2      4823671    4825122   \n",
       "MBURv2_30040          NaN  MBURv2       648480     649751   \n",
       "MBURv2_30115          int  MBURv2       749393     750415   \n",
       "MBURv2_60111         nirB  MBURv2      1491766    1494321   \n",
       "MBURv2_60113          NaN  MBURv2      1496503    1497081   \n",
       "MBURv2_60177          NaN  MBURv2      1561635    1561889   \n",
       "MBURv2_80064         glnG  MBURv2      1901015    1902424   \n",
       "MBURv2_80067         glnK  MBURv2      1906036    1906374   \n",
       "MBURv2_80070         amtB  MBURv2      1908452    1909744   \n",
       "\n",
       "                                                            note  translation  \\\n",
       "locus_tag                                                                       \n",
       "MBURv2_160283  Evidence 3 : Function proposed based on presen...          NaN   \n",
       "MBURv2_20180   Evidence 2a : Function of homologous gene expe...          NaN   \n",
       "MBURv2_20181   Evidence 2a : Function of homologous gene expe...          NaN   \n",
       "MBURv2_20183   Evidence 4 : Homologs of previously reported g...          NaN   \n",
       "MBURv2_210346                                                NaN          NaN   \n",
       "MBURv2_240071  Evidence 4 : Homologs of previously reported g...          NaN   \n",
       "MBURv2_240072  Evidence 4 : Homologs of previously reported g...          NaN   \n",
       "MBURv2_30040   Evidence 4 : Homologs of previously reported g...          NaN   \n",
       "MBURv2_30115                                                 NaN          NaN   \n",
       "MBURv2_60111   Evidence 2a : Function of homologous gene expe...          NaN   \n",
       "MBURv2_60113                                                 NaN          NaN   \n",
       "MBURv2_60177                                                 NaN          NaN   \n",
       "MBURv2_80064   Evidence 2a : Function of homologous gene expe...          NaN   \n",
       "MBURv2_80067   Evidence 2a : Function of homologous gene expe...          NaN   \n",
       "MBURv2_80070   Evidence 2a : Function of homologous gene expe...          NaN   \n",
       "\n",
       "               5GB1_FM40_T0m_TR2  5GB1_FM40_T10m_TR3  5GB1_FM40_T20m_TR2  \\\n",
       "locus_tag                                                                  \n",
       "MBURv2_160283         118.243568          112.275736          110.964306   \n",
       "MBURv2_20180          210.424105          158.632845          181.080545   \n",
       "MBURv2_20181           95.819838           86.744847           91.146440   \n",
       "MBURv2_20183           92.724781           77.558180           86.194789   \n",
       "MBURv2_210346          70.518396           68.822867           66.867096   \n",
       "MBURv2_240071         115.521215          125.266171          122.873477   \n",
       "MBURv2_240072         152.884221          166.004637          159.391668   \n",
       "MBURv2_30040          346.196665          456.748233          416.782502   \n",
       "MBURv2_30115           18.954208           19.738377           18.230891   \n",
       "MBURv2_60111          605.258423          691.437830          667.622137   \n",
       "MBURv2_60113           23.493069           27.461091           27.963444   \n",
       "MBURv2_60177          245.862088          217.642197          254.978512   \n",
       "MBURv2_80064          108.514855          116.924797          121.151452   \n",
       "MBURv2_80067         1236.258838         1480.016871         1362.990010   \n",
       "MBURv2_80070          178.796228          216.015070          211.287476   \n",
       "\n",
       "               5GB1_FM40_T40m_TR1  5GB1_FM40_T60m_TR1  5GB1_FM40_T90m_TR2  \\\n",
       "locus_tag                                                                   \n",
       "MBURv2_160283          100.836391           77.197657           70.032566   \n",
       "MBURv2_20180           126.342956           61.051857           70.194605   \n",
       "MBURv2_20181            71.621996           32.461622           39.674872   \n",
       "MBURv2_20183            70.352941           49.814896           47.463966   \n",
       "MBURv2_210346           63.493064           49.910520           54.874815   \n",
       "MBURv2_240071          171.859978          242.070964          230.034969   \n",
       "MBURv2_240072          244.378006          339.549453          333.040188   \n",
       "MBURv2_30040           555.641953         1007.956458          965.279974   \n",
       "MBURv2_30115            15.962341           11.687684           12.175099   \n",
       "MBURv2_60111           817.519655         1077.573416         1024.202207   \n",
       "MBURv2_60113            34.230125           49.110936           51.477168   \n",
       "MBURv2_60177           131.689317           45.544737           44.577813   \n",
       "MBURv2_80064           163.911171          229.124357          225.811603   \n",
       "MBURv2_80067          2351.761430         3215.524758         3305.708022   \n",
       "MBURv2_80070           293.525608          434.560734          446.470951   \n",
       "\n",
       "               5GB1_FM40_T150m_TR1_remake  5GB1_FM40_T180m_TR1  \n",
       "locus_tag                                                       \n",
       "MBURv2_160283                   73.893920            95.436505  \n",
       "MBURv2_20180                    95.464947           118.980515  \n",
       "MBURv2_20181                    55.479556            64.157192  \n",
       "MBURv2_20183                    55.260844            59.371024  \n",
       "MBURv2_210346                   53.845876            59.197819  \n",
       "MBURv2_240071                  215.806091           173.416287  \n",
       "MBURv2_240072                  298.420330           260.795367  \n",
       "MBURv2_30040                   855.040811           766.323303  \n",
       "MBURv2_30115                    14.009989            14.678014  \n",
       "MBURv2_60111                   988.430960           914.548557  \n",
       "MBURv2_60113                    44.846438            43.132774  \n",
       "MBURv2_60177                    81.965303           131.877279  \n",
       "MBURv2_80064                   183.639002           180.891342  \n",
       "MBURv2_80067                  3031.353754          2351.879330  \n",
       "MBURv2_80070                   376.033774           368.793380  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2_TPM.iloc[pearson_clusters[1],:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looks like there are two clusters, some expression and zero expression across samples."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "   ### Clustering mean centered euclidean distance with with HDBSCAN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "time to cluster 4.821575880050659\n"
     ]
    }
   ],
   "source": [
    "# Clustering the mean centered euclidean distance of TPM counts \n",
    "\n",
    "hdb_t1 = time.time()\n",
    "hdb_euclidean_mean = hdbscan.HDBSCAN(metric = \"precomputed\", min_cluster_size=10).fit(df3_euclidean_mean)\n",
    "hdb_euclidean_mean_labels = hdb_euclidean_mean.labels_\n",
    "hdb_elapsed_time = time.time() - hdb_t1\n",
    "print(\"time to cluster\", hdb_elapsed_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[-1]\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "print(np.unique(hdb_euclidean_mean_labels))\n",
    "print(np.bincount(hdb_euclidean_mean_labels[hdb_euclidean_mean_labels!=-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>product</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>locus</th>\n",
       "      <th>start_coord</th>\n",
       "      <th>end_coord</th>\n",
       "      <th>note</th>\n",
       "      <th>translation</th>\n",
       "      <th>5GB1_FM40_T0m_TR2</th>\n",
       "      <th>5GB1_FM40_T10m_TR3</th>\n",
       "      <th>5GB1_FM40_T20m_TR2</th>\n",
       "      <th>5GB1_FM40_T40m_TR1</th>\n",
       "      <th>5GB1_FM40_T60m_TR1</th>\n",
       "      <th>5GB1_FM40_T90m_TR2</th>\n",
       "      <th>5GB1_FM40_T150m_TR1_remake</th>\n",
       "      <th>5GB1_FM40_T180m_TR1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [product, type, gene_symbol, locus, start_coord, end_coord, note, translation, 5GB1_FM40_T0m_TR2, 5GB1_FM40_T10m_TR3, 5GB1_FM40_T20m_TR2, 5GB1_FM40_T40m_TR1, 5GB1_FM40_T60m_TR1, 5GB1_FM40_T90m_TR2, 5GB1_FM40_T150m_TR1_remake, 5GB1_FM40_T180m_TR1]\n",
       "Index: []"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "euclidean_mean_clusters = {i: np.where(hdb_euclidean_mean_labels == i)[0] for i in range(2)}\n",
    "df2_TPM.iloc[euclidean_mean_clusters[1],:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looks like 2 clusters - both with zero expression. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "looks like wether it is a numpy array or pandas dataframe, the result is the same. lets now try to get index of the clustered points. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "   ### Clustering log transformed euclidean distance with with HDBSCAN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "time to cluster 4.929334878921509\n"
     ]
    }
   ],
   "source": [
    "# Clustering the log2 transformed euclidean distance of TPM counts \n",
    "\n",
    "hdb_t1 = time.time()\n",
    "hdb_euclidean_log2 = hdbscan.HDBSCAN(metric = \"precomputed\", min_cluster_size=10).fit(df3_euclidean_log2)\n",
    "hdb_euclidean_log2_labels = hdb_euclidean_log2.labels_\n",
    "hdb_elapsed_time = time.time() - hdb_t1\n",
    "print(\"time to cluster\", hdb_elapsed_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[-1]\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "print(np.unique(hdb_euclidean_log2_labels))\n",
    "print(np.bincount(hdb_euclidean_log2_labels[hdb_euclidean_log2_labels!=-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>product</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>locus</th>\n",
       "      <th>start_coord</th>\n",
       "      <th>end_coord</th>\n",
       "      <th>note</th>\n",
       "      <th>translation</th>\n",
       "      <th>5GB1_FM40_T0m_TR2</th>\n",
       "      <th>5GB1_FM40_T10m_TR3</th>\n",
       "      <th>5GB1_FM40_T20m_TR2</th>\n",
       "      <th>5GB1_FM40_T40m_TR1</th>\n",
       "      <th>5GB1_FM40_T60m_TR1</th>\n",
       "      <th>5GB1_FM40_T90m_TR2</th>\n",
       "      <th>5GB1_FM40_T150m_TR1_remake</th>\n",
       "      <th>5GB1_FM40_T180m_TR1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [product, type, gene_symbol, locus, start_coord, end_coord, note, translation, 5GB1_FM40_T0m_TR2, 5GB1_FM40_T10m_TR3, 5GB1_FM40_T20m_TR2, 5GB1_FM40_T40m_TR1, 5GB1_FM40_T60m_TR1, 5GB1_FM40_T90m_TR2, 5GB1_FM40_T150m_TR1_remake, 5GB1_FM40_T180m_TR1]\n",
       "Index: []"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "euclidean_log2_clusters = {i: np.where(hdb_euclidean_log2_labels == i)[0] for i in range(2)}\n",
    "df2_TPM.iloc[euclidean_log2_clusters[1],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
